### ISSUE TITLE
Destroying Bastion host only, leads to all associated ec2 instances to be destroyed

### ISSUE THREAD

**Issue by TravSalesman** [terraform-providers/terraform-provider-aws#4246]
Hi everyone,

I have the following scenario, with which I stumbled across unexpected behavior:

**Description of the environment:**
Within my aws architecture that I deploy with terraform, I have a couple of ec2 instances that are public and also many that are private (i.e. no public ip address). I have some shell scripts and other config files in place that I provide to the instances through file provisioners, so that the ec2 instances are not spinned up with an initial configuration to automate the process of giving the specific instances their purpose.

Naturally, to achieve the provisioning of config files to be transferred to the instances on startup for the private instances, I must make use of a bastion host (which is an ec2 instance itself) as a forwarding proxy... which I do.
In my project I do intend to make use of the bastion host only for the purpose of the initial provisioning of the instances with their config files, and afterwards I have no use for the bastion host anymore (being more explicitly I want to remove the bastion host afterwards to minimize the attack surface on the resources within my VPC). Therefore I have placed the bastion host into a separate module, which I intend to destroy after the all resources were successfully deployed by terraform. In the chain of modules that are deployed by terraform, the bastion host gets initialized before the ec2 instances and I pass the data of the bastion host through output and input variables to the module that creates the ec2 instances. All instances receive their config files perfectly fine (as expected) when making use of the bastion host.

**TL;DR - Now to the problem:**
When I want to destroy the module that contains only the bastion host, the terraform plan accumulates that the private ec2 instances (that were dependend on the bastion host for initial provisioning of config files) must be destroyed too. 
As I've mentioned before, I only make use of the bastion host for initial setup and afterwards want to remove it from the network architecture, while I have to keep the ec2 instances alive. In the state as it is, I can no longer use the terraform script for further modifications and tracking of changes on the architecture, because terraform plan would want to destroy the private ec2 instances along with the bastion host, which itself would then again require the bastion host to have the private ec2 instances be setup properly again... i.e. endless circle of desctructions and initializations.

I basically want to achieve a state for my architecture, where all resources are properly setup and the bastion host has disappeared from the whole project. That way I could make further use of the terraform templates from that point on, to allow the architecture to evolve if necessary and still manage everything through terraform.

**Workarounds:**
What I could do of course is:

- Have the bastion host initialized separate from terraform and provide its data to the ec2 instances through parameters
- Have the bastion host initialized through an independent terraform call and destroy it the same way. But the bastion host itself has lots of dependencies on the network configuration of the main terraform project, so this would require lots of effort to get the correct input where it has to be.

Both ways do not fulfill the degree of integration and completeness within terraform, which I desire.

**Question:**
Therefore I'm wondering. Is there something I'm doing wrong (maybe regarding the _ignore_changes_ utility)? And if there is no option currently available to achieve what I'm looking for... would that be something that fits into the scope of functionality terraform shall provide or does it not meet the intended use of terraform (due to the fact that it is a two-stage process which I want to achieve).

I'm thankful for any hints!

### Terraform Version
v0.11.3

### Affected Resource(s)
aws_instance (with bastion host in use)

### Terraform Configuration Files
`
resource "aws_instance" "instance_A" {
 ...
 user_data = "${data.template_file.radius_server_user_data.rendered}"
 ...
  provisioner "file" {
    source      = "..."
    destination = "..."

    connection {
      bastion_host        = "${var.bastion_host_public_ip}"
      bastion_user        = "${var.bastion_host_user}"
      bastion_port        = "22"
      bastion_private_key = "${file("${var.bastion_host_ssh_key_file}")}"

      # activate usage of ssh-agent to allow forwarding of keys
      private_key = "${file("${var.radius_server_ssh_key_file}")}"
      user        = "..."
      type        = "ssh"
      agent       = true
    }
  }
lifecycle {
    ignore_changes = ["file.connection.bastion_host", "file.connection.bastion_user", "file.connection.bastion_port", "file.connection.bastion_private_key"]
  }
}
`

### Expected Behavior
When destroying the bastion host (by destroying the module that created the bastion host) after the terraform script has correctly initialized the aws instance, I would expect that terraform destroy would not want to destroy the aws instance that has made use of the bastion host. I tried to force that behavior through making use of the _ignore_changes_ utility, but it had no effect the way it is integrated into the code above.

### Actual Behavior
When destroying the bastion host (by destroying the module that created the bastion host) after the terraform script has correctly initialized the aws instance, the terraform plan intends to destroy all resources that made use of the bastion host (in this example the ec2 instance displayed in the code sample) for provisioning initial configuration.

### Steps to Reproduce
- create a bastion host
- pass the necessary data (e.g. ip address) of the bastion host via reference
- create an EC2 instance with the previously created bastion host in use for provisioning of initial configuration
- After terraform has created the mentioned resources correctly and provided the initial configuration, destroy the bastion host only (either the instance itself or the whole module containing it, in case it was placed in a separate module)

**Comment by mhaggerty042** [comment]
Facing the same issue. Another potential workaround could be:

1. Remove bastion host module's resources from your state using [`terraform state rm module.foo`](https://www.terraform.io/docs/commands/state/rm.html)
2. Manually delete bastion host resources from console

I also unsuccessfully tried to ignore the connection
```
resource "aws_instance" "example" {
   ...
   lifecycle {
       ignore_changes = ["connection"]
     }

     connection {
       type        = "ssh"
       user        = "ubuntu"
       private_key = "${file("~/.ssh/key.pem")}"

       bastion_host        = "${var.jumpbox_ip}"
       bastion_user        = "${var.jumpbox_user}"
       bastion_private_key = "${file("~/.ssh/${var.jumpbox_key}.pem")}"
     }
}
```

**Comment by TravSalesman** [comment]
Thanks a lot for the hint to remove the module from the state! To me it's the best workaround I found so far.

Also good idea to ignore the connection. Too bad it didn't work though. Of course, on the long run I'd say terraform should be able to properly handle the ignore_changes element in the described scenario.

**Comment by tracypholmes** [comment]
Thank you for using Terraform and for opening up this question! Issues on GitHub are intended to be related to bugs or feature requests with the provider codebase. 

It appears you've found a workaround, so this issue will now be closed. If needed, please use https://discuss.hashicorp.com/c/terraform-providers for additional feedback, community discussions, and questions around Terraform.

If you believe that your issue was closed in error, please create a new issue using one of the following provided templates: [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md) or [feature request](https://github.com/terraform-providers/terraform-provider-aws/issues/new?labels=enhancement&template=Feature_Request.md). Please make sure to provide us with the appropriate information so we can best determine how to assist with the given issue.
